radix-engine/src/blueprints/consensus_manager/consensus_manager.rs

```diff
@@ -446,20 +440,26 @@ impl ConsensusManagerBlueprint {
 
         Self::update_proposal_statistics(progressed_rounds as u64, proposal_history, api)?;
 
-        if round >= config.rounds_per_epoch {
-            let next_epoch = consensus_manager.epoch + 1;
-            Self::epoch_change(next_epoch, &config, api)?;
-            consensus_manager.epoch = next_epoch;
-            consensus_manager.round = 0;
+        let config = &config_substate.config;
+        let epoch_duration_millis = proposer_timestamp_milli - manager_substate.epoch_start_milli;
+        if config
+            .epoch_change_condition
+            .is_met(epoch_duration_millis, round)
+        {
+            let next_epoch = manager_substate.epoch + 1;
+            Self::epoch_change(next_epoch, config, api)?;
+            manager_substate.epoch = next_epoch;
+            manager_substate.epoch_start_milli = proposer_timestamp_milli;
```

Comment:
So this model will see some (slow) epoch drift over time, because they'll be a delay between when the 5 minutes is up and when a round change notices.

I wonder if a tiny bit of added complexity might be worth it to avoid this problem: we create separate fields `actual_epoch_start_milli` and `effective_epoch_start_milli` - and use `effective_epoch_start_milli` to measure `epoch_duration_millis`.

Then, when the epoch change condition is met, if `epoch_duration_millis` is sufficiently close to the epoch length (say, within 10%), then we set `manager_substate.effective_epoch_start_milli +=  standard_epoch_length` (and otherwise set it to `proposer_timestamp_milli` as currently).

This avoids the drift problem, as a slightly slow-to-change epoch will be offset in the next epoch :).

The `actual_epoch_start_milli` would just be as a sanity check and to let us monitor the drift per-epoch.

Reply:
I see the guaranteed drift.

However, is there an end-user value in mitigating it? To me, "not so great":
- for short-term epoch number estimations, that drift won't affect things that much.
- for long-term estimations (like, "what epoch number will we have on 2026.06.06 at 13:37?") the probability of `>10% diff => set it to proposer_timestamp_milli` is totally not negligible - which makes the estimation as unreliable as with a drift.

Let's not add this complexity just yet.